tango/workspaces/remote_workspace.py

```diff
@@ -0,0 +1,392 @@
+        # after the run.
+        self.client.upload(
+            run_dataset, json.dumps(run_data).encode(), self.Constants.RUN_DATA_FNAME
+        )
+
+        return Run(name=cast(str, name), steps=steps, start_date=run_dataset.created)
+
+    def registered_runs(self) -> Dict[str, Run]:
+        import concurrent.futures
+
+        runs: Dict[str, Run] = {}
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=9, thread_name_prefix="RemoteWorkspace.registered_runs()-"
+        ) as executor:
+            run_futures = []
+            for dataset in self.client.datasets(match=self.Constants.RUN_DATASET_PREFIX):
+                run_futures.append(executor.submit(self._get_run_from_dataset, dataset))
+            for future in concurrent.futures.as_completed(run_futures):
+                run = future.result()
+                if run is not None:
+                    runs[run.name] = run
+
+        return runs
+
+    def registered_run(self, name: str) -> Run:
+        err_msg = f"Run '{name}' not found in workspace"
+
+        try:
+            dataset_for_run = self.client.get(self.Constants.run_dataset_name(name))
+            # TODO: what's this check? seems to be beaker specific.
+            # # Make sure the run is in our workspace.
+            # if dataset_for_run.workspace_ref.id != self.beaker.workspace.get().id:
+            #     raise DatasetNotFound
+        except RemoteDatasetNotFound:
+            raise KeyError(err_msg)
+
+        run = self._get_run_from_dataset(dataset_for_run)
+        if run is None:
+            raise KeyError(err_msg)
+        else:
+            return run
+
+    @contextmanager
+    def capture_logs_for_run(self, name: str) -> Generator[None, None, None]:
+        with tempfile.TemporaryDirectory() as tmp_dir_name:
+            log_file = Path(tmp_dir_name) / "out.log"
+            try:
+                with file_handler(log_file):
+                    yield None
+            finally:
+                run_dataset = self.Constants.run_dataset_name(name)
+                self.client.sync(run_dataset, log_file)
+                self.client.commit(run_dataset)
+
+    def _get_run_from_dataset(self, dataset: RemoteDataset) -> Optional[Run]:
+        if dataset.name is None:
+            return None
+        if not dataset.name.startswith(self.Constants.RUN_DATASET_PREFIX):
+            return None
+
+        try:
+            run_name = dataset.name[len(self.Constants.RUN_DATASET_PREFIX) :]
+            steps: Dict[str, StepInfo] = {}
+            steps_info_bytes = self.client.get_file(dataset, self.Constants.RUN_DATA_FNAME)
+            steps_info = json.loads(steps_info_bytes)
+        except (RemoteDatasetNotFound, FileNotFoundError):
+            return None
+
+        import concurrent.futures
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=9, thread_name_prefix="RemoteWorkspace._get_run_from_dataset()-"
```

Comment:
This number 9 is carefully chosen to not overwhelm Beaker. It should not be hardcoded here. Google can likely handle more traffic.